keyboards/bpiphany/ghost_squid/matrix.c

```diff
@@ -0,0 +1,167 @@
+{
+  for (uint8_t col = 0; col < MATRIX_COLS; col++) {
+    select_row(col);
+    wait_us(30);
+    matrix_row_t rows = read_cols();
+    for (uint8_t row = 0; row < MATRIX_ROWS; row++) {
+      bool prev_bit = matrix_debouncing[row] & ((matrix_row_t)1<<col);
+      bool curr_bit = rows & (1<<row);
+      if (prev_bit != curr_bit) {
+	matrix_debouncing[row] ^= (matrix_row_t) 1 << col;
+	debouncing = DEBOUNCE;
+      }
+    }
+  }
+
+  if (debouncing) {
+    if (--debouncing) {
+      wait_ms(1);
+    } else {
+      for (uint8_t i = 0; i < MATRIX_ROWS; i++) {
+	matrix[i] = matrix_debouncing[i];
+      }
+    }
+  }
+
+  matrix_scan_quantum();
+
+  return 1;
+}
+
+bool matrix_is_modified(void)
+{
+  if (debouncing)
+    return false;
+  return true;
+}
+
+inline
+bool matrix_is_on(uint8_t row, uint8_t col)
+{
+  return matrix[row] & 1 << col;
+}
+
+inline
+matrix_row_t matrix_get_row(uint8_t row)
+{
+  return matrix[row];
+}
+
+void matrix_print(void)
+{
+  print("\nr/c 0123456789ABCDEF\n");
+  for (uint8_t row = 0; row < MATRIX_ROWS; row++) {
+    print_hex4(row); print(": ");
+    print_bin_reverse16(matrix_get_row(row));
+    print("\n");
+  }
+}
+
+uint8_t matrix_key_count(void)
+{
+  uint8_t count = 0;
+  for (uint8_t i = 0; i < MATRIX_ROWS; i++) {
+    count += bitpop16(matrix[i]);
+  }
+  return count;
+}
+
+static matrix_row_t read_cols(void)
+{
+  return
+    (PINB&(1<<1) ? 0 : ((matrix_row_t)1<<0)) |
+    (PINC&(1<<2) ? 0 : ((matrix_row_t)1<<1)) |
+    (PINB&(1<<6) ? 0 : ((matrix_row_t)1<<2)) |
+    (PINB&(1<<4) ? 0 : ((matrix_row_t)1<<3)) |
+    (PINB&(1<<3) ? 0 : ((matrix_row_t)1<<4)) |
+    (PINB&(1<<2) ? 0 : ((matrix_row_t)1<<5)) |
+    (PINB&(1<<0) ? 0 : ((matrix_row_t)1<<6)) |
+    (PINB&(1<<5) ? 0 : ((matrix_row_t)1<<7)); 
+}
+
+static void select_row(uint8_t col)
+{
+  switch (col) {
+  case  0: PORTD = (PORTD & ~0b01111110) | 0b01100010; break;
+  case  1: PORTD = (PORTD & ~0b01111110) | 0b01101000; break;
+  case  2: PORTD = (PORTD & ~0b01111110) | 0b01101100; break;
+  case  3: PORTD = (PORTD & ~0b01111110) | 0b01110000; break;
+  case  4: PORTD = (PORTD & ~0b01111110) | 0b01111000; break;
+  case  5: PORTD = (PORTD & ~0b01111110) | 0b01100000; break;
+  case  6: PORTD = (PORTD & ~0b01111110) | 0b01110100; break;
+  case  7: PORTD = (PORTD & ~0b01111110) | 0b01100100; break;
+  case  8: PORTD = (PORTD & ~0b01111110) | 0b01111100; break;
+  case  9: PORTD = (PORTD & ~0b01111110) | 0b01101010; break;
+  case 10: PORTD = (PORTD & ~0b01111110) | 0b00110110; break;
+  case 11: PORTD = (PORTD & ~0b01111110) | 0b00010110; break;
+  case 12: PORTD = (PORTD & ~0b01111110) | 0b01001110; break;
+  case 13: PORTD = (PORTD & ~0b01111110) | 0b00111110; break;
+  case 14: PORTD = (PORTD & ~0b01111110) | 0b00011110; break;
+  case 15: PORTD = (PORTD & ~0b01111110) | 0b01000110; break;
+  case 16: PORTD = (PORTD & ~0b01111110) | 0b00100110; break;
+  case 17: PORTD = (PORTD & ~0b01111110) | 0b00101110; break;
+  }
+}
```

Comment:
```suggestion
matrix_row_t read_rows(void) {
    return
        (PINB & (1 << 1) ? 0 : ((matrix_row_t)1 << 0)) |
        (PINC & (1 << 2) ? 0 : ((matrix_row_t)1 << 1)) |
        (PINB & (1 << 6) ? 0 : ((matrix_row_t)1 << 2)) |
        (PINB & (1 << 4) ? 0 : ((matrix_row_t)1 << 3)) |
        (PINB & (1 << 3) ? 0 : ((matrix_row_t)1 << 4)) |
        (PINB & (1 << 2) ? 0 : ((matrix_row_t)1 << 5)) |
        (PINB & (1 << 0) ? 0 : ((matrix_row_t)1 << 6)) |
        (PINB & (1 << 5) ? 0 : ((matrix_row_t)1 << 7));
}

void select_col(uint8_t col) {
    switch (col) {
        case  0: PORTD = (PORTD & ~0b01111110) | 0b01100010; break;
        case  1: PORTD = (PORTD & ~0b01111110) | 0b01101000; break;
        case  2: PORTD = (PORTD & ~0b01111110) | 0b01101100; break;
        case  3: PORTD = (PORTD & ~0b01111110) | 0b01110000; break;
        case  4: PORTD = (PORTD & ~0b01111110) | 0b01111000; break;
        case  5: PORTD = (PORTD & ~0b01111110) | 0b01100000; break;
        case  6: PORTD = (PORTD & ~0b01111110) | 0b01110100; break;
        case  7: PORTD = (PORTD & ~0b01111110) | 0b01100100; break;
        case  8: PORTD = (PORTD & ~0b01111110) | 0b01111100; break;
        case  9: PORTD = (PORTD & ~0b01111110) | 0b01101010; break;
        case 10: PORTD = (PORTD & ~0b01111110) | 0b00110110; break;
        case 11: PORTD = (PORTD & ~0b01111110) | 0b00010110; break;
        case 12: PORTD = (PORTD & ~0b01111110) | 0b01001110; break;
        case 13: PORTD = (PORTD & ~0b01111110) | 0b00111110; break;
        case 14: PORTD = (PORTD & ~0b01111110) | 0b00011110; break;
        case 15: PORTD = (PORTD & ~0b01111110) | 0b01000110; break;
        case 16: PORTD = (PORTD & ~0b01111110) | 0b00100110; break;
        case 17: PORTD = (PORTD & ~0b01111110) | 0b00101110; break;
    }
}

void matrix_init_custom(void) {
    /* Column output pins */
    setPinOutput(D1);
    setPinOutput(D2);
    setPinOutput(D3);
    setPinOutput(D4);
    setPinOutput(D5);
    setPinOutput(D6);

    /* Row input pins */
    writePinHigh(B0);
    writePinHigh(B1);
    writePinHigh(B2);
    writePinHigh(B3);
    writePinHigh(B4);
    writePinHigh(B5);
    writePinHigh(B6);
    writePinHigh(C2);
}

bool matrix_scan(matrix_row_t current_matrix[]) {
    bool changed = false;

    for (uint8_t col = 0; col < MATRIX_COLS; col++) {
        select_col(col);
        matrix_io_delay();
        matrix_row_t rows = read_rows();

        for (uint8_t row = 0; row < MATRIX_ROWS; row++) {
            bool prev_bit = current_matrix[row] & ((matrix_row_t)1 << col);
            bool curr_bit = rows & (1 << row);

            if (prev_bit != curr_bit) {
                current_matrix[row] ^= (matrix_row_t)1 << col;
                changed = true;
            }
        }
    }

    return changed;
}
```
Not 100% sure about this, but might work. There's still some stuff to be cleaned up.